### ISSUE TITLE
redis.asyncio.Redis is very slow compared redis.Redis

### ISSUE THREAD

**Issue by vardanaloyan** [redis/redis-py#3412]
Asynchronous redis client performs several (hundreds, thousands) time slow than sync client, in similar conditions.

**Version**: redis==5.0.6

**Platform**: python:3.12.4-slim-bookworm docker image

**Description**: 

I use redis in async web server application, which has certain response time requirements.
After using async client, I noticed significant drop of the response time.
I have prepared small script that could also show-case the problem and some numbers.
 
```python
import asyncio
import json
import time

import uvloop
from redis import Redis
from redis.asyncio import Redis as RedisAsync

loop = uvloop.new_event_loop()
asyncio.set_event_loop(loop)


async def measure_sync(redis_proxy):
    st = time.perf_counter()
    redis_proxy.get("data")
    return time.perf_counter() - st


async def measure_async(redis_proxy):
    st = time.perf_counter()
    await redis_proxy.get("data")
    return time.perf_counter() - st


async def main():
    N = 1000
    host = "v2-redis"
    port = 6379
    db1 = 10
    db2 = 11

    initial_data = {
        "language": "Python",
    }

    redis_proxy_sync = Redis(host=host, port=port, db=db1, decode_responses=True)
    redis_proxy_async = RedisAsync(host=host, port=port, db=db2, decode_responses=True)

    redis_proxy_sync.set("data", json.dumps(initial_data))
    await redis_proxy_async.set("data", json.dumps(initial_data))

    tasks_with_sync_redis = [measure_sync(redis_proxy_sync) for _ in range(N)]
    tasks_with_async_redis = [measure_async(redis_proxy_async) for _ in range(N)]
    sync_execution_times = await asyncio.gather(*tasks_with_sync_redis)
    async_execution_times = await asyncio.gather(*tasks_with_async_redis)
    avg_sync_execution_time = sum(sync_execution_times) / N
    avg_async_execution_time = sum(async_execution_times) / N
    print(f"Average sync execution time: {1000*avg_sync_execution_time:.6f} milliseconds")
    print(f"Average async execution time: {1000*avg_async_execution_time:.6f} milliseconds")
    print(
        f"The sync version is faster than the async version {avg_async_execution_time/avg_sync_execution_time:.2f} times !"
    )


if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(main())

```

After execution I have following results:

```
Average sync execution time: 0.051713 milliseconds
Average async execution time: 90.637792 milliseconds
The sync version is faster than the async version 1752.72 times !
```

**Comment by mazen-isstech** [comment]
This could be due to how the tasks are concurrently executed in:
```python
async_execution_times = await asyncio.gather(*tasks_with_async_redis)
```
What likely happened was that all tasks began by first executing:
```python
st = time.perf_counter()
```
Then concurrently, for every task, ran:
```python
await redis_proxy.get("data")
```
Finally, one by one, they exited the call above and dropped into:
```python
return time.perf_counter() - st
```
Which caused every `time.perf_counter() - st` to not just count the time for a single Redis Get call, but all of them, approximately.

Something to note about asyncio is that it will not run your calls in parallel, but rather have the tasks frequently return control to the event loop, creating an illusion of parallelism. Asyncio has at least two benefits, 1. It is fair. Requests are processed concurrently (but not in parallel). 2. If you have a heavily IO-bound operations, asyncio can relieve the application from waiting for the operation to end, and give it an opportunity to process other tasks until results are back. A simple Redis Get call is likely to be very fast in terms of waiting for results from the network. So imo, asyncio wont be able truly "shine" in this example.

tl;dr I suspect that it's not a fair comparison.

**Comment by vardanaloyan** [comment]
Thanks for the response. As I mentioned above, I have a working application which is async web server (based on asyncio + uvloop) and I noticed a significant drop on the performance when I switch using async client instead of sync. After doing some profiling and in-depth analysis I understood, that root cause of the problem is async redis client, and in order to validate it I've created the script which I've shared here.

So the aim was not to showcase that async is worst than sync, and for that reason I've chosen the worst specific case to showcase that. 
The aim was just to provide some example that show cases what I have noticed in my application. (Obliviously I couldn't share application code here).

And in my application I don't have any kind of time.perfcounters but still performance is several times slow than sync redis.

If you have some better ideas how it could be illustrated then please share it with me as well.

**Comment by mazen-isstech** [comment]
Execute the tests sequentially. 

Replace:
```python
    tasks_with_sync_redis = [measure_sync(redis_proxy_sync) for _ in range(N)]
    tasks_with_async_redis = [measure_async(redis_proxy_async) for _ in range(N)]
    sync_execution_times = await asyncio.gather(*tasks_with_sync_redis)
    async_execution_times = await asyncio.gather(*tasks_with_async_redis)
```
With:
```python
    sync_execution_times = [await measure_sync(redis_proxy_sync) for _ in range(N)]
    async_execution_times = [await measure_async(redis_proxy_async) for _ in range(N)]
```

Here are some results on my side.

N=1000:
```
Average sync execution time: 0.040341 milliseconds
Average async execution time: 0.055501 milliseconds
The sync version is faster than the async version 1.38 times !
```
N=10000:
```
Average sync execution time: 0.042801 milliseconds
Average async execution time: 0.056712 milliseconds
The sync version is faster than the async version 1.33 times !
```

There is a noticeable overhead of using asyncio, but it's definitely not that dramatic.

I want to reemphasize the fact that Redis Get is a relatively cheap operation. Asyncio will help especially when a read from the network blocks for a longer time. In that case, a similar test would show a dramatic difference, but in favor of asyncio.

**Comment by vardanaloyan** [comment]
In the sequential example it make sense, but I'm afraid that it's going into another direction from my use case. 
I don't have sequential N calls in real application, but the web server is being called parallel by many clients, that was the reason that I've used asyncio.gather, trying to simulate the "load" which I have in real environment.

Maybe `asyncio.gather` is not the correct way to simulate the "load". 

So for the moment, by using async client application faces significant drop of the response time around 10 times, which is not acceptable for us.

**Comment by mazen-isstech** [comment]
Replacing:
```python
    tasks_with_sync_redis = [measure_sync(redis_proxy_sync) for _ in range(N)]
    tasks_with_async_redis = [measure_async(redis_proxy_async) for _ in range(N)]
    sync_execution_times = await asyncio.gather(*tasks_with_sync_redis)
    async_execution_times = await asyncio.gather(*tasks_with_async_redis)
    avg_sync_execution_time = sum(sync_execution_times) / N
    avg_async_execution_time = sum(async_execution_times) / N
```

With:
```python
    now = time.perf_counter()
    for _ in range(N):
        redis_proxy_sync.get("data")
    sync_execution_time = time.perf_counter() - now

    now = time.perf_counter()
    await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])
    async_execution_time = time.perf_counter() - now

    avg_sync_execution_time = sync_execution_time / N
    avg_async_execution_time = async_execution_time / N
```

Gave this result (N=100):
```
Average sync execution time: 0.083145 milliseconds
Average async execution time: 8.077742 milliseconds
The sync version is faster than the async version 97.15 times !
```

Yeah, that looks terrible.

**Comment by mazen-isstech** [comment]
I figured something out, internally the Redis client has a connection pool. If you run multiple calls concurrently, the pool will establish a connection in the pool for every call. We have been measuring the time it takes to establish a connection + the Redis Get operation.

The fix in this benchmark is to force Redis to establish N connections in the pool by running `asyncio.gather` before measuring the calls.

Try this instead:
```python
    redis_proxy_sync.set("data", json.dumps(initial_data))
    await redis_proxy_async.set("data", json.dumps(initial_data))
    await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])

    now = time.perf_counter()
    for _ in range(N):
        redis_proxy_sync.get("data")
    sync_execution_time = time.perf_counter() - now

    now = time.perf_counter()
    await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])
    async_execution_time = time.perf_counter() - now

    avg_sync_execution_time = sync_execution_time / N
    avg_async_execution_time = async_execution_time / N
```

Result (N=1000):
```
Average sync execution time: 0.073929 milliseconds
Average async execution time: 0.050592 milliseconds
The sync version is faster than the async version 0.68 times !
```

**Comment by vardanaloyan** [comment]
It's misleading. By putting `time.perf_counter`s directly in `main` function, you execute them when you run the in the bottom of the file by this line `asyncio.get_event_loop().run_until_complete(main())`. 

So the sync redis execution time is correct, but async's time not. Because by executing `main()` it's not doing the job yet. It should be awaited still.

**Comment by vardanaloyan** [comment]
Well, thanks for the snippet, I'll have to check deeper in the application code and get back here.

**Comment by vardanaloyan** [comment]
So sync client does not have connection pool and the behavior and async has ? I'm trying to understand the core issue, why we have to initiate     `await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])` ? In real application this is something that cannot be the fix I guess.

**Comment by mazen-isstech** [comment]
Both have connection pools. But the sync one does not need to create more than one connection because all operations happen sequentially.
If you add after the line you mentioned:
```python
    print("Sync connections:",
          len(redis_proxy_sync.connection_pool._available_connections))
    print("Async connections:",
          len(redis_proxy_async.connection_pool._available_connections))
```
You should get this output:
```
Sync connections: 1
Async connections: 1000
```

> In real application this is something that cannot be the fix I guess.

I would agree. But at least it illustrates the issue.

**Comment by vardanaloyan** [comment]
I've done the changes related creating connections in advance, and it solved the issue in application also providing much better and stable results than sync client was giving. 
What would be better, that redis python library supported dedicated method for "warming up" the connections in the connection pool, rather than doing `await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])`. Although that is different topic. For the moment I'll close the issue. Thanks a lot for digging into the problem and finding a root cause.

**Comment by mazen-isstech** [comment]
You're welcome.

**Comment by big-camel** [comment]
> I've done the changes related creating connections in advance, and it solved the issue in application also providing much better and stable results than sync client was giving. What would be better, that redis python library supported dedicated method for "warming up" the connections in the connection pool, rather than doing `await asyncio.gather(*[redis_proxy_async.get("data") for _ in range(N)])`. Although that is different topic. For the moment I'll close the issue. Thanks a lot for digging into the problem and finding a root cause.

I encountered the same issue, and now I'm using the following method to pre-create the required number of connections. Is there a better approach?

```py
async def pre_create_connections(self, n: int):
        async def create_connection(connection_pool: redis.ConnectionPool):  
            connection = connection_pool.make_connection()
            await connection_pool.ensure_connection(connection)
            connection_pool._available_connections.append(connection)
        tasks = [create_connection(self._connection_pool) for _ in range(n)]
        await asyncio.gather(*tasks)
```

**Comment by markwitt1** [comment]
Aby update on this? Does it make sense to use sync redis if we have a lot of different tasks in asyncio at the same time?
